src/envoy/http/authn/authn_utils.cc

```diff
@@ -0,0 +1,97 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "authn_utils.h"
+#include "common/common/base64.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+
+bool AuthnUtils::GetJWTPayloadFromHeaders(
+    const HeaderMap& headers, const LowerCaseString& jwt_payload_key,
+    ::google::protobuf::Map< ::std::string, ::std::string>* claims) {
+  const HeaderEntry* entry = headers.get(jwt_payload_key);
+  if (!entry) {
+    ENVOY_LOG(error, "{}: no JwtPayloadKey entry {} in the header",
+              __FUNCTION__, jwt_payload_key.get());
+    return false;
+  }
+  std::string value(entry->value().c_str(), entry->value().size());
+  std::string payload_str = Base64::decode(value);
+  // Return an empty string if Base64 decode fails.
+  if (payload_str.empty()) {
+    ENVOY_LOG(error, "Invalid {} header, invalid base64: {}",
+              jwt_payload_key.get(), value);
+    return false;
+  }
+  try {
+    auto json_obj = Json::Factory::loadFromString(payload_str);
+    ENVOY_LOG(debug, "{}: json object is {}", __FUNCTION__,
+              json_obj->asJsonString());
+    json_obj->iterate(
+        [claims](const std::string& key, const Json::Object& obj) -> bool {
+          ENVOY_LOG(debug, "{}: key is {}", __FUNCTION__, key);
+          try {
+            if (key == "aud") {
+              // "aud" can be either string array or string.
+              // Try as string array, read it as empty array if doesn't exist.
+              std::vector<std::string> aud_vector;
+              try {
+                aud_vector = obj.getStringArray(key, true);
+              } catch (Json::Exception& e) {
+                // Try as string
+                try {
+                  auto audience = obj.asString();
+                  aud_vector.push_back(audience);
+                } catch (Json::Exception& e) {
+                  ENVOY_LOG(error,
+                            "{}: an exception is thrown when get audience as "
+                            "string. The exception is: {}",
+                            __FUNCTION__, e.what());
+                }
+              }
+              std::string aud_str;
+              // If aud_vector has more than 1 items, the result audiences are
+              // delimited by ','
+              for (size_t i = 0; i < aud_vector.size(); i++) {
```

Comment:
Why do you want to concat aud data? It may destroy the usability of that field later.
On a closer look at the purpose of this function, it's better that you combine the buildAttribute in to this function, and make function output JwtPayload directly.
For those special attribute (iss, aud, presenter), write them to the designate location, for the rest, put into claims map (it's ok to ignore non string claims for now)

Reply:
Done.